src/app/services/shared-state.service.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * @overview 
+ * The state is a shared data store that provides some blocks with asynchronous access to data. 
+ * This capability should never modify the data emitted by the block, as this could easily result in infinite loops within out flows. 
+ * It can however be used in elements of the UI to present up to date information to the user, or to enable/disable components. 
+ * 
+ * To add a shared state capabilities to a component, you need to subscribe to updates in the constructor. 
+ * 
+ * @example
+ *   constructor(
+ *    private stateService:SharedStateService
+ *    ) {
+ *        stateService.shared$.subscribe(state => { setTimeout(() =>{this.updateOutputDisplay()}) });
+ *       }
+ * 
+ *  The setTimeout() wrapper allows the updates to work with Angular's change detection cycle. Don't leave it out. 
```

Comment:
I'm curious, without setTimeout, what happens? Are ExpressionChangedAfterItHasBeenCheckedError errors raised?
I'm not an Angular expect, but I wonder if a better solution would be to do the subscription at a different stage of the component's lifecycle (if possible?) e.g: ngAfterContentChecked or ngAfterContentInit vs in the constructor?
@the-greenman

Reply:
Yes, the setTimeout deals with the ExpressionChangedAfterItHasBeenCheckedError. 

I'll see if there are other possible subscription locations.

Reply:
Doesn't look like the location of the subscribe matters. It just sets up the listener, which will then get called when data arrives.

Reply:
Hmm. Maybe I'm lying. ngAfterViewInit is working in my initial test

Reply:
nope.

Reply:
I've attempted to use the onPush change detection strategy. And then call "this.cd.markForCheck()" when I have an update. But still get the issue. 

I've tried the subscribe in multiple locations.

Reply:
https://hackernoon.com/everything-you-need-to-know-about-the-expressionchangedafterithasbeencheckederror-error-e3fd9ce7dbb4

Here is a handy writeup. I think the underlying challenge is that our flows are using the main angular change detection. So, the "correct" way to do this is to pass the changes down through the flow again.  But this is not what we want. 

The solution that looks viable from this article (one of many I've read) seems to be to create a microtask with a promise, rather than a macrotask with the timeout.

Reply:
Cool ok I assume it runs a tiny bit faster with a promise than a 0ms setTimeout then.

> I've attempted to use the onPush change detection strategy. And then call "this.cd.markForCheck()" when I have an update. But still get the issue.
> 
> I've tried the subscribe in multiple locations.

Interesting, can it work with some subset of blocks without defering in any way?
If this is too big a change to make to existing blocks, this could be tech debt we should accept for now, and improve on one piece at a time.